File: lib/AbstractLayer/AbsLayerDid.cpp
//
//  Did.cpp
//  lib
//
//  Created by mengxk on 2019/3/23.
//  Copyright © 2019 Elastos. All rights reserved.
//

#include "AbsLayerDid.hpp"

#include <Did.h>
#include <WalletError.h>

char* AbsLayer_Did_GetId(int64_t obj)
{
  std::shared_ptr<elastos::Did>* did = reinterpret_cast<std::shared_ptr<elastos::Did>*>(obj);
  std::string id = (*did)->GetId();
  char* idBuf = strdup(id.c_str());
  return idBuf;
}

char* AbsLayer_Did_SignInfo(int64_t obj, const char* seed, const char* json)
{
  std::shared_ptr<elastos::Did>* did = reinterpret_cast<std::shared_ptr<elastos::Did>*>(obj);
  std::string signedInfo = (*did)->SignInfo(seed, json);
  char* signedInfoBuf = strdup(signedInfo.c_str());
  return signedInfoBuf;
}

char* AbsLayer_Did_SetInfo(int64_t obj, const char* seed, const char* json, int64_t wallet)
{
  std::shared_ptr<elastos::Did>* did = reinterpret_cast<std::shared_ptr<elastos::Did>*>(obj);
  std::shared_ptr<elastos::HDWallet>* hdWallet = reinterpret_cast<std::shared_ptr<elastos::HDWallet>*>(wallet);
  std::string txid = (*did)->SetInfo(seed, json, *hdWallet);
  char* txidBuf = strdup(txid.c_str());
  return txidBuf;
}

int AbsLayer_Did_SyncInfo(int64_t obj)
{
  if (obj == 0) {
    return E_WALLET_C_INVALID_ARGUMENT;
  }
  std::shared_ptr<elastos::Did>* did = reinterpret_cast<std::shared_ptr<elastos::Did>*>(obj);
  int ret = (*did)->SyncInfo();
  return ret;
}

char* AbsLayer_Did_GetInfo(int64_t obj, const char* key)
{
  if (obj == 0) {
    return nullptr;
  }
  
  std::shared_ptr<elastos::Did>* did = reinterpret_cast<std::shared_ptr<elastos::Did>*>(obj);
  std::string value = (*did)->GetInfo(key);
  char* valueBuf = strdup(value.c_str());
  return valueBuf;
}

void AbsLayer_Did_DestroyDid(int64_t obj)
{
  if (obj == 0) {
    return;
  }
  std::shared_ptr<elastos::Did>* did = reinterpret_cast<std::shared_ptr<elastos::Did>*>(obj);
  did->reset();
  delete did;
}

File: lib/AbstractLayer/AbsLayerDid.swift
//
//  AbsLayerDid.swift
//  lib
//
//  Created by mengxk on 2019/3/23.
//  Copyright © 2019 Elastos. All rights reserved.
//

extension AbstractLayer {
  final class Did {
    private init() {}
    
    public static func GetId(obj: Int64) -> String? {
      let idPtr = AbsLayer_Did_GetId(obj)
      let id = String.FromUnsafeMutablePointer(data: idPtr)
      AbstractLayer_FreeBuf(idPtr)
      return id
    }
    
    public static func SignInfo(obj: Int64, seed: String, json: String) -> String? {
      let seedPtr = String.ToUnsafeMutablePointer(data: seed)
      let jsonPtr = String.ToUnsafeMutablePointer(data: json)
      let signedInfoPtr = AbsLayer_Did_SignInfo(obj, seedPtr, jsonPtr)
      let signedInfo = String.FromUnsafeMutablePointer(data: signedInfoPtr)
      AbstractLayer_FreeBuf(signedInfoPtr)
      return signedInfo
    }
    
    public static func SetInfo(obj: Int64, seed: String, json: String, wallet: Int64) -> String? {
      let seedPtr = String.ToUnsafeMutablePointer(data: seed)
      let jsonPtr = String.ToUnsafeMutablePointer(data: json)
      let txidPtr = AbsLayer_Did_SetInfo(obj, seedPtr, jsonPtr, wallet)
      let txid = String.FromUnsafeMutablePointer(data: txidPtr)
      AbstractLayer_FreeBuf(txidPtr)
      return txid
    }
    
    public static func SyncInfo(obj: Int64) {
      AbsLayer_Did_SyncInfo(obj)
    }
    
    public static func GetInfo(obj: Int64, key: String) -> String? {
      let keyPtr = String.ToUnsafeMutablePointer(data: key)
      let infoPtr = AbsLayer_Did_GetInfo(obj, keyPtr)
      let info = String.FromUnsafeMutablePointer(data: infoPtr)
      AbstractLayer_FreeBuf(infoPtr)
      return info
    }
    
    public static func DestroyDid(obj: Int64) {
      AbsLayer_Did_DestroyDid(obj)
    }
  }
}



